### ISSUE TITLE
BlobService class blobs with "question marks" in the path issue

### ISSUE THREAD

**Issue by huguesv** [Azure/azure-storage-python#67]
Moved from https://github.com/Azure/azure-sdk-for-python/issues/494

Hi guys,

We are using Azure storage with one of our clients, and we are using the python sdk.

When we yield over the container_name with list_blobs method we get blobs with "question marks".
 After that if we try to get the same blob thru the get_blob_to_file (or get_blob) method,
 We get "Not found (The specified blob does not exist.)" exception.

I see that the issue is in the "_update_request_uri_query" method.
 You guys expect a query string after "question marks" but you allowing blobs with "question marks" in the path...

For example:
 if my path is: "test_path/test name?.txt"
 I see that in the request the path is "//test_path/test name"

B.T.W:
 I see that in the C# sdk it was fixed
http://stackoverflow.com/questions/12242120/azure-blob-reference-to-file-with-question-mark-in-the-name

please advice,

Thanks,
 Lior
li.mizr@gmail.com

**Comment by huguesv** [comment]
I was able to repro the issue.  The http request for list blobs operation is not returning the correct name for the blob (it's missing the question mark).  See code to repro below.
```
from azure.storage.blob import BlobService

s = BlobService(n, k)

container_name = 'questionmarktest'
blob_name = 'hello?'

s.create_container('questionmarktest')
s.put_block_blob_from_text(container_name, blob_name, 'hello world')

text = s.get_blob_to_text(container_name, blob_name)
print(text)

for blob in s.list_blobs(container_name):
    # this prints out 'hello' instead of the expected 'hello?'
    print(blob.name)

# response from list_blobs http get:
# <?xml version="1.0" encoding="utf-8"?><EnumerationResults ServiceEndpoint="https://ptvsazureunittests.blob.core.windows.net/" ContainerName="questionmarktest"><Blobs><Blob><Name>hello</Name><Properties><Last-Modified>Fri, 30 Oct 2015 21:16:34 GMT</Last-Modified><Etag>0x8D2E16F6394E7B5</Etag><Content-Length>11</Content-Length><Content-Type>application/octet-stream Charset=UTF-8</Content-Type><Content-Encoding /><Content-Language /><Content-MD5>XrY7u+Ae7tCTyyK7j1rNww==</Content-MD5><Cache-Control /><Content-Disposition /><BlobType>BlockBlob</BlobType><LeaseStatus>unlocked</LeaseStatus><LeaseState>available</LeaseState></Properties></Blob></Blobs><NextMarker /></EnumerationResults>
```

**Comment by emgerner-msft** [comment]
The repro does work and this is a bug, but the guilty code here actually has nothing to do with list_blobs. If you actually look at the request being sent for the blob creation the library isn't including the '?' in the blob name. list_blobs just reveals that problem.

The code causing a problem here is the _update_request_uri_query method in _common_serialization. We use the '?' to delimit the query parameter section of the URL and do not appropriately differentiate if it was passed in intentionally as part of the blob name. The functionality right now would be if you did a blob name something like hello?foo=bla, we'd send hello as the blob name and foo=bla as a query param in the url. This isn't really useful and probably wasn't intentional so we should fix. However, as the request path we send to this method frequently does include query parameters (see create_container code for an example) this isn't a trivial change and may take some time.

**Comment by jehine-MSFT** [comment]
Hello @liormizr & @alfpark ,

We have released the 0.30 version (#132) which addressed this feature request. If you have issues with the new version in regards to this issue, please reopen it and continue the discussion.

Thanks!
